### PR TITLE
Expose Additional Callbacks

### LINKED ISSUE #1051: Remove DownloadAdapter
Remove `DownloadAdapter` and have `GetDownloadHandler` implemented in `ClientAdapter`, resolves the problem of passing in a new instance of `CefSharpBrowserWrapper`.

(Also it's consistent with the other handlers).

### PR DESCRIPTION
**Pull request by amaitland**
- Resolve #1051 
- Add CefDownloadItemCallback Wrapper

### CODE REVIEW

**Comment by amaitland** [comment]
Decided to expose the additional callbacks to have proper `async` support.

Still need to add

- [ ] `CefAuthCallback`
- [ ] `CefFileDialogCallback`
- [ ] `CefGeolocationCallback`
- [ ] `CefBeforeDownloadCallback`

File: CefSharp/IDownloadHandler.cs [hunk 1]
```diff
@@ -11,17 +11,17 @@ public interface IDownloadHandler
         /// </summary>
         /// <param name="browser">The browser instance</param>
         /// <param name="downloadItem">Represents the file being downloaded.</param>
-        /// <param name="downloadPath">Path where the file will be saved if <see cref="showDialog"/> is False.</param>
-        /// <param name="showDialog">Display a dialog allowing the user to specify a custom path and filename.</param>
+        /// <param name="callback">Callback interface used to asynchronously continue a download.</param>
         /// <returns>Return True to continue the download otherwise return False to cancel the download</returns>
-        bool OnBeforeDownload(IBrowser browser, DownloadItem downloadItem, out string downloadPath, out bool showDialog);
+        void OnBeforeDownload(IBrowser browser, DownloadItem downloadItem, IBeforeDownloadCallback callback);
```
Comment (rassilon) [inline CefSharp/IDownloadHandler.cs]: Need to make a pass through and remove the now incorrect return XML comments on these methods.

File: CefSharp/IDownloadHandler.cs [hunk 2]
```diff
@@ -11,17 +11,17 @@ public interface IDownloadHandler
...
 
         /// <summary>
         /// Called when a download's status or progress information has been updated. This may be called multiple times before and after <see cref="OnBeforeDownload"/>.
         /// </summary>
         /// <param name="browser">The browser instance</param>
         /// <param name="downloadItem">Represents the file being downloaded.</param>
+        /// <param name="callback">The callback used to Cancel/Pause/Resume the process</param>
         /// <returns>Return True to cancel, otherwise False to allow the download to continue.</returns>
```
Comment (rassilon) [inline CefSharp/IDownloadHandler.cs]: Like here

File: CefSharp/IGeolocationHandler.cs [hunk 3]
```diff
@@ -12,8 +12,9 @@ public interface IGeolocationHandler
         /// <param name="browser">the browser object</param>
         /// <param name="requestingUrl">the URL requesting permission</param>
         /// <param name="requestId">the unique ID for the permission request</param>
+        /// <param name="callback">Callback interface used for asynchronous continuation of geolocation permission requests.</param>
         /// <returns>true to allow the request and false to deny</returns>
-        bool OnRequestGeolocationPermission(IWebBrowser browser, string requestingUrl, int requestId);
+        bool OnRequestGeolocationPermission(IWebBrowser browser, string requestingUrl, int requestId, IGeolocationCallback callback);
```
Comment (rassilon) [inline CefSharp/IGeolocationHandler.cs]: Shouldn't this be void? (I haven't looked at the CEF interface).

File: CefSharp/IDownloadHandler.cs (hunk 1 above)
Comment (amaitland) [inline CefSharp/IDownloadHandler.cs]: :+1:

File: CefSharp/IGeolocationHandler.cs (hunk 3 above)
Comment (amaitland) [inline CefSharp/IGeolocationHandler.cs]: `CEF API` returns `bool` (Would have made sense to be `void`)

**Comment by amaitland** [comment]
Expanded `ISchemeHandlerFactory.Create` to pass in new `IBrowser`, `IFrame`, `IRequest` wrappers, plus `schemeName`

Now you can actually load a different handler for based on different params.

**Comment by amaitland** [comment]
Merged directly into `master`

**Pull request closed without merging**